k3k-kubelet/provider/provider.go

```diff
@@ -193,75 +221,223 @@ func (p *Provider) PortForward(ctx context.Context, namespace, pod string, port
 
 // CreatePod takes a Kubernetes Pod and deploys it within the provider.
 func (p *Provider) CreatePod(ctx context.Context, pod *corev1.Pod) error {
-	translated := p.translateTo(pod)
+	tPod := pod.DeepCopy()
+	p.Translater.TranslateTo(tPod)
+
 	// these values shouldn't be set on create
-	translated.UID = ""
-	translated.ResourceVersion = ""
-	// can't handle volumes yet, just omit them for now
-	translated.Spec.Volumes = nil
-	containers := []corev1.Container{}
-	for _, container := range translated.Spec.Containers {
-		container.VolumeMounts = nil
-		containers = append(containers, container)
-	}
-	translated.Spec.Containers = containers
-	translated.Spec.NodeName = ""
-	p.logger.Infof("Creating pod %s/%s for pod %s/%s", translated.Namespace, translated.Name, pod.Namespace, pod.Name)
-	_, err := p.CoreClient.Pods(p.ClusterNamespace).Create(ctx, translated, metav1.CreateOptions{})
-	return err
+	tPod.UID = ""
+	tPod.ResourceVersion = ""
+
+	// the node was scheduled on the virtual kubelet, but leaving it this way will make it pending indefinitely
+	tPod.Spec.NodeName = ""
+
+	// volumes will often refer to resources in the virtual cluster, but instead need to refer to the sync'd
+	// host cluster version
+	err := p.transformVolumes(ctx, pod.Namespace, tPod.Spec.Volumes)
+	if err != nil {
+		return fmt.Errorf("unable to sync volumes for pod %s/%s: %w", pod.Namespace, pod.Name, err)
+	}
+	p.logger.Infof("Creating pod %s/%s for pod %s/%s", tPod.Namespace, tPod.Name, pod.Namespace, pod.Name)
+	return p.HostClient.Create(ctx, tPod)
+}
+
+// transformVolumes changes the volumes to the representation in the host cluster. Will return an error
+// if one/more volumes couldn't be transformed
+func (p *Provider) transformVolumes(ctx context.Context, podNamespace string, volumes []corev1.Volume) error {
+	for _, volume := range volumes {
+		// note: this needs to handle downward api volumes as well, but more thought is needed on how to do that
+		if volume.ConfigMap != nil {
+			if err := p.syncConfigmap(ctx, podNamespace, volume.ConfigMap.Name); err != nil {
+				return fmt.Errorf("unable to sync configmap volume %s: %w", volume.Name, err)
```

Comment:
Not really sure what is the correct way here but since we are using zap all around maybe we can use zap.Fields instead of fmt.Errorf, better actually if the logger can be passed to the transformVolumes function and add the fields there so that it can actually appear what kind of resources are we working on, as an example:

https://github.com/rancher/k3k/blob/main/pkg/controller/cluster/pod.go#L65-L66

Reply:
I don't have a problem with things near log statement using zap, but I think that errors in the program shouldn't be coupled tightly to the logger in case we want to change it.

Reply:
Like I said above, I can definitely change this when logging, but when returning an error internally we should avoid reliance on unnecessary third party libraries. I can make a structured log statement here in addition to the error return if you would prefer.